src/dune_rules/blang.ml

```diff
@@ -57,6 +57,96 @@ let rec eval t ~dir ~f =
     and y = String_with_vars.expand y ~mode:Many ~dir ~f in
     Op.eval op (Value.L.compare_vals ~dir x y)
 
+module Partial = struct
+  type partial_value =
+    | Unexpanded of String_with_vars.t
+    | Expanded of Value.t list
+
+  type t =
+    | Const of bool
+    | Expr of String_with_vars.t
+    | And of t list
+    | Or of t list
+    | Compare of Op.t * partial_value * partial_value
+
+  let of_bool b = Const b
+
+  let to_bool = function
+    | Const b -> Some b
+    | _ -> None
+
+  let rec eval t ~dir ~f =
+    match t with
+    | Const x -> x
+    | Expr sw -> (
+      match String_with_vars.expand sw ~mode:Single ~dir ~f with
+      | String "true" -> true
+      | String "false" -> false
+      | _ ->
+        let loc = String_with_vars.loc sw in
+        User_error.raise ~loc
+          [ Pp.text "This value must be either true or false" ] )
+    | And xs -> List.for_all ~f:(eval ~f ~dir) xs
+    | Or xs -> List.exists ~f:(eval ~f ~dir) xs
+    | Compare (op, x, y) ->
+      let expand = function
+        | Expanded l -> l
+        | Unexpanded v -> String_with_vars.expand v ~mode:Many ~dir ~f
+      in
+      let x = expand x
+      and y = expand y in
+      Op.eval op (Value.L.compare_vals ~dir x y)
+end
+
+let rec partial_eval t ~dir ~f =
+  match t with
+  | Const x -> Partial.Const x
+  | Expr sw -> (
+    match String_with_vars.partial_expand sw ~mode:Single ~dir ~f with
+    | Expanded (String "true") -> Partial.Const true
+    | Expanded (String "false") -> Partial.Const false
+    | Expanded _ ->
+      let loc = String_with_vars.loc sw in
+      User_error.raise ~loc
+        [ Pp.text "This value must be either true or false" ]
+    | Unexpanded sw -> Partial.Expr sw )
+  | And xs ->
+    let rec aux acc = function
+      | [] -> (
+        match acc with
+        | [] -> Partial.Const true
+        | acc -> Partial.And (List.rev acc) )
+      | a :: l -> (
+        match partial_eval a ~dir ~f with
+        | Partial.Const true -> aux acc l
+        | Partial.Const false as r -> r
+        | r -> aux (r :: acc) l )
+    in
+    aux [] xs
+  | Or xs ->
```

Comment:
The `And` and `Or` cases are non-trivial and almost a copy of each other. Could you refactor the code to avoid duplication of logic?